homeassistant/components/modbus/__init__.py

```diff
@@ -178,73 +146,136 @@ def write_coil(service):
+            loop, client = ClientUDP(
+                schedulers.ASYNC_IO,
+                host=self._config_host,
+                port=self._config_port,
+                timeout=self._config_timeout,
+                loop=self._loop,
+            )
+        else:
+            assert False
+        self._client = client.protocol
+        _LOGGER.debug("setup done with loop %s", loop)
 
     def close(self):
         """Disconnect client."""
-        with self._lock:
-            self._client.close()
-
-    def connect(self):
-        """Connect client."""
-        with self._lock:
-            self._client.connect()
+        self._client.close()
 
-    def read_coils(self, unit, address, count):
+    async def read_coils(self, unit, address, count):
         """Read coils."""
-        with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_coils(address, count, **kwargs)
+        await self._connect_delay()
+        kwargs = {"unit": unit} if unit else {}
+        return await self._client.read_coils(address, count, **kwargs)
 
-    def read_discrete_inputs(self, unit, address, count):
+    async def read_discrete_inputs(self, unit, address, count):
         """Read discrete inputs."""
-        with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_discrete_inputs(address, count, **kwargs)
+        await self._connect_delay()
+        kwargs = {"unit": unit} if unit else {}
+        return await self._client.read_discrete_inputs(address, count, **kwargs)
 
-    def read_input_registers(self, unit, address, count):
+    async def read_input_registers(self, unit, address, count):
         """Read input registers."""
-        with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_input_registers(address, count, **kwargs)
+        await self._connect_delay()
+        kwargs = {"unit": unit} if unit else {}
+        return await self._client.read_input_registers(address, count, **kwargs)
 
-    def read_holding_registers(self, unit, address, count):
+    async def read_holding_registers(self, unit, address, count):
         """Read holding registers."""
-        with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            return self._client.read_holding_registers(address, count, **kwargs)
+        await self._connect_delay()
+        kwargs = {"unit": unit} if unit else {}
+        return await self._client.read_holding_registers(address, count, **kwargs)
 
     def write_coil(self, unit, address, value):
         """Write coil."""
-        with self._lock:
-            kwargs = {"unit": unit} if unit else {}
-            self._client.write_coil(address, value, **kwargs)
+        kwargs = {"unit": unit} if unit else {}
+        self._client.write_coil(address, value, **kwargs)
 
     def write_register(self, unit, address, value):
```

Comment:
The writes will need to be converted to async as well like `def write_coil(service):`

Reply:
done